### PR TITLE
Add ReduxReducer extension

### PR DESCRIPTION
**Pull request by vojtechszocs**
This PR adds new `ReduxReducer` extension.

Example usage taken from KubeVirt plugin:

```ts
// plugin.tsx
{
  type: 'ReduxReducer',
  properties: {
    namespace: 'kubevirt',
    reducer: kubevirtReducer, // (SubState, Action) => SubState
    required: FLAG_KUBEVIRT,
  },
},
```

This allows Console plugins to add new Redux reducers which operate on (are scoped to)

```
plugins.<namespace>
````

sub-state within the `RootState` object.

The `required` property can be used to gate such extensions by feature flag(s).

The `namespace` property must be unique across all plugins. This is enforced via dedicated extension check.

Helpers in `public/components/dashboard/utils` were moved to plugin registry:
- `isDashboardExtensionInUse` :arrow_right: `registry.isExtensionInUse`
- `getFlagsForExtensions` :arrow_right: `registry.getRequiredFlags`

Finally, this new extension is used in KubeVirt plugin to replace the current VM wizard implementation, reducing the coupling to `kubevirt-web-ui-components` library.

@rawagner @suomiy I've moved the main KubeVirt reducer to `kubevirt-plugin/src/redux`.

@jelkosz This also turns on the new VM wizard implementation by @suomiy.

### CODE REVIEW

**Comment by vojtechszocs** [comment]
@spadgett @rawagner 

Marking as WIP since the new VM wizard implementation needs to be tested more thoroughly.

**Comment by vojtechszocs** [comment]
@rawagner An alternative would be to make the VM wizard use [`useReducer`](https://reactjs.org/docs/hooks-reference.html#usereducer) hook.

**Comment by rawagner** [comment]
> @rawagner An alternative would be to make the VM wizard use [`useReducer`](https://reactjs.org/docs/hooks-reference.html#usereducer) hook.

that would be very difficult to do especially with given timeframe and @suomiy on PTO. Not really an option

**Comment by rawagner** [comment]
@vojtechszocs tested the wizard and it works. Can you please remove WIP ?

**Comment by vojtechszocs** [comment]
> @vojtechszocs tested the wizard and it works. Can you please remove WIP ?

Done, @spadgett @christianvogt can you please share your thoughts?

We have a React component in KubeVirt plugin which would like to manage some of its data within the app Redux (sub)state via plugin-specific reducer function, is this OK?

**Comment by rawagner** [comment]
rebased on latest master

**Comment by spadgett** [comment]
I think we probably need something like this for plugins to be use Redux. I don't see us moving away from Redux anytime soon even with hooks. My concern is this is coming in pretty late.

@christianvogt opinion?

**Comment by jelkosz** [comment]
> My concern is this is coming in pretty late.

its not that late, it is before the feature freeze meaning way before the code freeze. If something breaks, there is plenty of time to fix it, or, in worst case, revert it. At least this is how I understand the feature freeze milestone - until the feature freeze, the features can get in (even large ones). After that, they can not...

**Comment by christianvogt** [comment]
@spadgett I to see value in a redux extension. So conceptually I'm ok with it.

I question the use of redux in the wizard component though and think hooks w/ context would be better to isolate the component. Unless there's some unique need for redux that I'm unaware of.

From a time perspective I have no problem with it.

**Comment by jelkosz** [comment]
Can we maybe separate the discussion? It seems that both you and Sam agree that this kind of an extension has value. This PR is about adding it. So lets just review the pr in this context. 

It is true that it is needed by one dialog in the kubevirt plugin but that is a separate topic. I would really hate to see us missing the freeze with our new and shiny pf4 vm wizard and staying with a pf3 modal dialog especially considering there is no fundamental disagreement here.

**Comment by spadgett** [comment]
I'm OK with the change as long as we can get it reviewed and merged by feature freeze

/approve

File: frontend/public/redux.ts [hunk 1]
```diff
@@ -33,17 +36,59 @@ export type RootState = {
...
+    const featureState = store.getState()[featureReducerName];
+    return featureState ? _.pick(featureState.toObject(), requiredFlags) : null;
+  };
+
+  let flagsObject = getReduxFlagsObject();
+
+  store.subscribe(() => {
+    const currentFlagsObject = getReduxFlagsObject();
+
+    if (JSON.stringify(flagsObject) !== JSON.stringify(currentFlagsObject)) {
```
Comment (spadgett) [inline frontend/public/redux.ts]: I think this is probably OK, but it relies on the order of keys in `JSON.stringify` to be stable. It's not guaranteed, but I believe it happens in practice.

**Comment by vojtechszocs** [comment]
Thanks guys for your comments & review.

Will rebase.

@spadgett I agree that relying on key order in `JSON.stringify` isn't that good, I can change it to

```ts
_.isEqual(flagsObject, currentFlagsObject)
```

which should have the same effect but not depend on enumerated key order.

**Comment by spadgett** [comment]
We should be mindful of performance here as this check will run a lot. I think flags is a simple flat object, so deep equals is OK. I wonder if we can get away with `===` or is it a different object every time?

**Comment by spadgett** [comment]
/lgtm cancel

needs rebase

File: frontend/public/redux.ts (hunk 1 above)
Comment (vojtechszocs) [reply frontend/public/redux.ts]: > We should be mindful of performance here as this check will run a lot.

:+1:

Every time the Console Redux state changes, the new listener gets called, so we narrow the data-to-compare (`getReduxFlagsObject`) down to flags required by _any_ `ReduxReducer` extensions.

```ts
const getReduxFlagsObject = () => {
  const requiredFlags = registry.getRequiredFlags([isReduxReducer]);
  const featureState = store.getState()[featureReducerName]; // <= all flags
  return featureState ? _.pick(featureState.toObject(), requiredFlags) : null; <= narrow flags
};
```

With current code, there is 1x instance of `ReduxReducer` extension (contributed by KubeVirt plugin) gated by the `KUBEVIRT` flag, so `getReduxFlagsObject` will return an object with _only_ `KUBEVIRT` key, at the moment.

> I think flags is a simple flat object, so deep equals is OK. I wonder if we can get away with `===` or is it a different object every time?

It's a different object upon every `getReduxFlagsObject` call due to `_.pick`.

**Comment by vojtechszocs** [comment]
> We should be mindful of performance here as this check will run a lot.

:+1:

Every time the Console Redux state changes, the new listener gets called, so we narrow the data to compare (`getReduxFlagsObject`) down to only flags required by _any_ `ReduxReducer` extensions.

```ts
const getReduxFlagsObject = () => {
  const requiredFlags = registry.getRequiredFlags([isReduxReducer]);
  const featureState = store.getState()[featureReducerName]; // <= all flags
  return featureState ? _.pick(featureState.toObject(), requiredFlags) : null; <= narrow flags
};
```

With current code, there is 1x instance of `ReduxReducer` extension (contributed by KubeVirt plugin) gated by the `KUBEVIRT` flag, so `getReduxFlagsObject` will return an object with _only_ `KUBEVIRT` key.

> I think flags is a simple flat object, so deep equals is OK. I wonder if we can get away with `===` or is it a different object every time?

It's a different object upon every `getReduxFlagsObject` call due to `_.pick`.

**Comment by vojtechszocs** [comment]
> I question the use of redux in the wizard component though and think hooks w/ context would be better to isolate the component. Unless there's some unique need for redux that I'm unaware of.

I can follow-up on this once @suomiy is back, but I _think_ it has to do with the fact that there can be multiple wizard component instances that share some state - notice `createVmWizards` (plural form) in `packages/kubevirt-plugin/src/utils/redux/reducers.ts`.

**Comment by vojtechszocs** [comment]
PR updated - comparison changed to use `_.isEqual` and `@console/plugin-sdk` imports moved above relative imports.

**Comment by spadgett** [comment]
> its not that late, it is before the feature freeze meaning way before the code freeze.

My concern is that it's fundamental code that runs on *every* Redux state change in the app and has the potential to break things or introduce basic performance problems outside the single wizard where it's used. I feel like one day before feature freeze is pretty late. We only have one more sprint to find and fix problems.

**Comment by mareklibra** [comment]
> I question the use of redux in the wizard component though and think hooks w/ context would be better to isolate the component. Unless there's some unique need for redux that I'm unaware of.

As written in another comment, Redux is here and there are no plans to remove it.
However, if this assumption is changed, we can reuse the reducers (mostly).

I agree that the `useReducer()` would isolate the top-level component, which is strong argument. But most probably can lock the use of subcomponents to just the tree of the one particular use case (it would add extra effort to avoid that).

One of the benefits of Redux is in no need to use `useContext()` to pass the state down. The use of `connect()` simplifies the code when the hierarchy is complex as in the case of `Create VM Wizard`. When writing the "subcomponents", we focus on `props` only.

The `useReducer()` hook is not meant to be replacement of Redux. Anyway, the decision between these two is not black-and-white.

The `Crete VM Wizard` is complex like a small application and the complexity is expected to further grow, the Wizard is not fully finished yet. That was the reason to consider state management using reducers. For less-complex components, the use of reducers would be anti-pattern, imo.

> Can we maybe separate the discussion? It seems that both you and Sam agree that this kind of an extension has value. This PR is about adding it. So lets just review the pr in this context.

Agree, considering time-frame.

**Comment by rawagner** [comment]
rebased

**Comment by jelkosz** [comment]
/lgtm

Adding back label after rebase

**Comment by jelkosz** [comment]
Come on CI, you can do it!

/retest

**Pull request merged**